src/orion/core/io/cmdline_parser.py

```diff
@@ -0,0 +1,229 @@
+
+    Parameters
+    ----------
+    parsing_tokens : dict, optional
+        Dictionary containing the tokens to use when parsing `-` and `_`.
+
+    Attributes
+    ----------
+    arguments : OrderedDict
+        Commandline arguments' name and value(s).
+    template : list
+        List of template-ready strings (in the form 'something_{0}') for formatting.
+
+    """
+
+    def __init__(self, parsing_tokens=None):
+        """See `CmdlineParser` description"""
+        self.arguments = OrderedDict()
+        self._already_parsed = False
+        self.template = []
+
+        parsing_tokens = {'-': '??', '_': '!!'}
+
+        if parsing_tokens is not None:
+            parsing_tokens.update(parsing_tokens)
+
+        self._dash_token = parsing_tokens['-']
+        self._underscore_token = parsing_tokens['_']
+
+    def format(self, configuration):
+        """Format the current template.
+
+        Recreate the string of argument using the template made at parse-time and
+        the values inside the `configuration` argument.
+
+        Parameters
+        ----------
+        configuration : dict
+            Dictionary storing the keys and values to be passed to the `format` function.
+            This would typically be `parser.arguments` where `parser` is a `CmdlineParser`
+            instance.
+
+        Returns
+        -------
+        str
+            A recreated string of the commandline passed to Orion.
+
+        """
+        return " ".join(self.template).format(**configuration)
+
+    def parse(self, commandline):
+        """Parse the `commandline` argument.
+
+        Create an OrderedDict where the keys are the names of the arguments and the values
+        are the actual values of the each argument. The arguments can be a single value or a
+        list of values. This also supports positional arguments.
+
+        Parameters
+        ----------
+        commandline : list
+            List of string representing the commmandline arguments.
+
+        Returns
+        -------
+        OrderedDict
+            Dictionary holding the values of every argument. The keys are the arguments' name.
+
+        Raises
+        ------
+        ValueError
+            If there is a duplicate argument
+
+        Notes
+        -----
+        By default, all values are `str` unless their types have been changed in the meantime.
+        Unnamed arguments's key follow the format '_X' where `X` is the index of that argument
+        in the list. For example, `val1 val2` would be parsed as
+        `{'_pos_0': 'val1', '_pos_1': 'val2}`.
+
+        Arrays are parsed starting at the first named argument until the next one. For example :
+        `--arg1 value1 value2 --arg2 value3 value4` will be parsed as :
+        `{'arg1': ['value1', 'value2'], 'arg2': ['value3', 'value4']}`.
+
+        File paths are extended to their absolute forms.
+
+        If the commandline contains optional arguments before a subcommand, the arguments
```

Comment:
Should turn this into a [warning section](https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#warnings)

Reply:
Noted.